pulsar-broker/src/main/java/org/apache/pulsar/compaction/TwoPhaseCompactor.java

```diff
@@ -271,7 +266,7 @@ private void phaseTwoLoop(RawReader reader, MessageId to, Map<String, MessageId>
                                         promise.completeExceptionally(exception2);
                                     }
                                 });
-                        if (to.equals(id)) {
+                        if (lastReadId.equals(id)) {
```

Comment:
There seems to be a return missing in line-275, may need to open another PR to fix it.

Reply:
Yes, if we don't stop phase two, the next entry may be dropped because it does not exist in the `latestForKey`.

Reply:
Another PR will fix it, please see https://github.com/apache/pulsar/pull/20988